### PR TITLE
#1399. [Records] Type inference and promotion tests added

### PR DESCRIPTION
**Pull request by sgrekhov**
I believe that this set of tests covers promotion and inference for records. If I missed anything and we need more tests, please, let me know

### CODE REVIEW

**Comment by sgrekhov** [comment]
@eernstg please review

File: LanguageFeatures/Records/type_inference_A01_t01.dart [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+// SharedOptions=--enable-experiment=records
+
+import "../../Utils/static_type_helper.dart";
+
+main() {
+  var r1 = (42, "");
+  r1.expectStaticType<Exactly<(int, String)>>();
+
+  final r2 = (null, 3.14);
+  r2.expectStaticType<Exactly<(Null, double)>>();
```
Comment (eernstg) [inline LanguageFeatures/Records/type_inference_A01_t01.dart:21]: This is a very interesting case! We have the following:

```dart
void main() {
  var x = null;
  String s = x; // No compile-time error.
}
```

The point is that a variable of type `Null` isn't very useful, so type inference chooses `dynamic` as the declared type when a variable initializer has the type `Null`.

But I can see that we get `Null` as an inferred type argument for normal instance creation inference:

```dart
class C<X, Y> {
  final X x;
  final Y y;
  C(this.x, this.y);
}

void main() {
  var c = C(null, 3.14);
  print(c.runtimeType); // 'C(Null, double)'.
}
```

With that in mind I think it's OK to expect that `null` gives rise to an inferred record field type of `Null`.

File: LanguageFeatures/Records/type_promotion_A01_t03.dart [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    r1.$0;
+    r1.s;
+  } else {
+    r1?.$0;
+    r1?.s;
+  }
+
+  R2 r2 = (42, o: "answer");
+  if (r2 is (int? n, {String o})) {
+    r2.n?.isOdd;
```
Comment (eernstg) [inline LanguageFeatures/Records/type_promotion_A01_t03.dart:33]:
```suggestion
    r2.$0?.isOdd;
```
We can't use a name associated with a positional field as a getter on the record.

**Review by eernstg — COMMENTED** [review]
Looks good!

One thing would be useful: In some cases there is an if condition (in `if (b) ...`, that's `b`) which is expected to evaluate to false, and we only get the static analysis of the body of the if-statement, because it won't be executed at run time. In this case it would be good to have a different record to perform the test on, such that the condition is true and the code gets executed.

(This is certainly true for the cases where a test is performed on `Record.empty`, but I think I noticed a couple of other cases along the way.)

**Comment by sgrekhov** [comment]
Done. Please rereview

**Review by eernstg — APPROVED** [review]
LGTM!

**Pull request merged**